File: services/python/app/models/content.py
"""内容库：scenarios / scenario_messages / songs / lrc / song_pitch_refs / listening_materials。

写归属（docs/10 §3）：
- ``scenarios`` / ``songs`` / ``lrc`` / ``listening_materials`` —— **Java（内容 CRUD + 上下架）**；
- ``scenario_messages``（会话消息流水，对话运行时产生）—— **Python**；
- ``song_pitch_refs``（离线参考旋律提取结果，算法侧产出）—— **Python**；
  与 Java 主表 songs/lrc 的耦合靠「lrc_id 外键 + 内容下架/删除联动」解耦（见 docs/10 §3.2）。
"""

from __future__ import annotations

from decimal import Decimal

from sqlalchemy import (
    BigInteger,
    CheckConstraint,
    ForeignKey,
    Index,
    Numeric,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from .base import (
    Base,
    ContentStatus,
    CreatedAtMixin,
    MessageRoles,
    PitchRefStatus,
    TimestampMixin,
    bigint_pk,
    jsonb,
)

# 场景类型（docs/06 §9.6 定稿 4 类 + other 兜底）
SCENE_TYPES = ("cafe", "airport", "interview", "library", "other")

# 内容来源（商用音乐严禁入库，docs/06 §9.7）
CONTENT_SOURCES = ("public_domain", "original", "demo_only")


class Scenario(TimestampMixin, Base):
    """对话场景模板（Java 写）。内容：角色设定 / 开场白 / 目标语料 / 难度分级。

    - prompt_version：system_prompt 修订序号。会话中 LLM 每轮记快照，
      旧会话可在报告/复盘中按版本解释行为差异（不改历史）；
    - interest_tags 给推荐系统做匹配；status 仅 draft/published/archived（不物理删除）。
    """

    __tablename__ = "scenarios"

    id: Mapped[int] = bigint_pk()
    title: Mapped[str] = mapped_column(String(128), nullable=False)
    scene_type: Mapped[str] = mapped_column(String(32), nullable=False)
    difficulty: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    description: Mapped[str | None] = mapped_column(String(512))
    system_prompt: Mapped[str] = mapped_column(Text, nullable=False)  # 角色设定
    opening_line: Mapped[str] = mapped_column(Text, nullable=False)  # 开场白
    target_corpus: Mapped[str | None] = mapped_column(Text)  # 目标语料/句式
    interest_tags: Mapped[dict] = mapped_column(
        jsonb(), nullable=False, server_default=text("'[]'")
    )
    prompt_version: Mapped[int] = mapped_column(
        SmallInteger, nullable=False, server_default=text("1")
    )
    estimated_turns: Mapped[int | None] = mapped_column(SmallInteger)
    estimated_minutes: Mapped[int | None] = mapped_column(SmallInteger)
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text(f"'{ContentStatus.DRAFT}'")
    )

    __table_args__ = (
        CheckConstraint(
            "scene_type IN ('cafe', 'airport', 'interview', 'library', 'other')", name="scene_type"
        ),
        CheckConstraint("difficulty BETWEEN 1 AND 4", name="difficulty"),
        CheckConstraint(
            f"status IN ('{ContentStatus.DRAFT}', '{ContentStatus.PUBLISHED}', "
            f"'{ContentStatus.ARCHIVED}')",
            name="status",
        ),
        # 推荐/列表页常用过滤组合
        Index("ix_scenarios_status_diff", "status", "difficulty"),
        Index("ix_scenarios_scene_type", "scene_type"),
    )


class ScenarioMessage(CreatedAtMixin, Base):
    """会话消息流水（Python 写；不可变，只 INSERT）。

    字段设计支撑 docs/06 §9.1 指标口径：
    - ``origin``：user 消息区分「主动发起(proactive)/响应(respond)」——互动率分子分母；
    - ``action``：功能互动（点示范/纠错/二次录制）——互动率补充指标；
    - ``prompt_version``：该轮 LLM 使用 system prompt 版本快照；
    - ``meta``：轮次级流水线状态（ASR/评分/LLM/TTS 各步耗时与错误快照，排障用）。
    """

    __tablename__ = "scenario_messages"

    id: Mapped[int] = bigint_pk()
    session_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("sessions.id", ondelete="CASCADE"), nullable=False
    )
    seq: Mapped[int] = mapped_column(SmallInteger, nullable=False)  # 会话内序号，从 1
    role: Mapped[str] = mapped_column(String(16), nullable=False)
    origin: Mapped[str | None] = mapped_column(String(24))
    action: Mapped[str | None] = mapped_column(String(16))
    content: Mapped[str] = mapped_column(Text, nullable=False)
    audio_url: Mapped[str | None] = mapped_column(String(512))
    duration_ms: Mapped[int | None] = mapped_column(BigInteger)
    prompt_version: Mapped[int | None] = mapped_column(SmallInteger)
    meta: Mapped[dict] = mapped_column(jsonb(), nullable=False, server_default=text("'{}'"))

    __table_args__ = (
        UniqueConstraint("session_id", "seq", name="uq_scenario_messages_session_seq"),
        CheckConstraint(
            f"role IN ('{MessageRoles.SYSTEM}', '{MessageRoles.USER}', '{MessageRoles.ASSISTANT}')",
            name="role",
        ),
        CheckConstraint(
            "origin IS NULL OR (role = 'user' AND origin IN ('proactive', 'respond'))",
            name="origin",
        ),
        CheckConstraint(
            "action IN ('demo', 'correction', 'retry', 'hint') OR action IS NULL", name="action"
        ),
    )


class Song(TimestampMixin, Base):
    """歌曲库（Java 写）。demo 只用公有领域/自创曲，商用音乐不入库（docs/06 §9.7）。

    - duration_s/bpm/key 供节奏对齐粗校准（文档层面，实际以 song_pitch_refs 为准）；
    - lrc_url 仅存原始 LRC 文件引用（可选）；逐句数据在 lrc 表（逐句评分唯一真源）。
    """

    __tablename__ = "songs"

    id: Mapped[int] = bigint_pk()
    title: Mapped[str] = mapped_column(String(128), nullable=False)
    artist: Mapped[str | None] = mapped_column(String(128))
    level: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    duration_s: Mapped[int | None] = mapped_column(BigInteger)
    bpm: Mapped[Decimal | None] = mapped_column(Numeric(6, 2))
    musical_key: Mapped[str | None] = mapped_column(String(8))
    audio_url: Mapped[str] = mapped_column(String(512), nullable=False)
    lrc_url: Mapped[str | None] = mapped_column(String(512))
    cover_url: Mapped[str | None] = mapped_column(String(512))
    interest_tags: Mapped[dict] = mapped_column(
        jsonb(), nullable=False, server_default=text("'[]'")
    )
    source: Mapped[str] = mapped_column(
        String(32), nullable=False, server_default=text("'public_domain'")
    )
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text(f"'{ContentStatus.DRAFT}'")
    )
    # 参考旋律就绪状态（docs/11 Q-B11）：LRC 重写→song_pitch_refs 级联清→离线重提取期间
    # 跟唱请求见 status!='ready' 返回「生成中/缺词」，不静默算分；由 Python 离线任务翻转
    pitch_ref_status: Mapped[str] = mapped_column(
        String(24), nullable=False, server_default=text(f"'{PitchRefStatus.MISSING}'")
    )

    __table_args__ = (
        CheckConstraint("level BETWEEN 1 AND 4", name="level"),
        CheckConstraint(
            f"status IN ('{ContentStatus.DRAFT}', '{ContentStatus.PUBLISHED}', "
            f"'{ContentStatus.ARCHIVED}')",
            name="status",
        ),
        CheckConstraint("source IN ('public_domain', 'original', 'demo_only')", name="source"),
        CheckConstraint(
            f"pitch_ref_status IN ('{PitchRefStatus.MISSING}', '{PitchRefStatus.BUILDING}', "
            f"'{PitchRefStatus.READY}', '{PitchRefStatus.INVALID}')",
            name="pitch_ref_status",
        ),
        Index("ix_songs_status_level", "status", "level"),
    )


class Lrc(CreatedAtMixin, Base):
    """逐句歌词（Java 写）。逐句评分唯一真源。

    - offset_ms/end_offset_ms：句窗口（end 为 docs/10 设计补充，评分对齐用；
      若 Java 沿用纯 LRC 无 end，可空，由算法按下一句 offset 推断）；
    - 编辑方式为「整首重写」：改 LRC → 删旧插新（seq 1..n 重排），
      song_pitch_refs 随 lrc_id 级联删除并离线重提取（Python 侧）。
    """

    __tablename__ = "lrc"

    id: Mapped[int] = bigint_pk()
    song_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("songs.id", ondelete="CASCADE"), nullable=False
    )
    seq: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    offset_ms: Mapped[int] = mapped_column(BigInteger, nullable=False)
    end_offset_ms: Mapped[int | None] = mapped_column(BigInteger)
    # 注意：属性名用 line_text（列名仍是 text）——若属性命名为 text 会遮蔽模块级
    # text() 函数，同 class body 内后续 server_default 全部崩（docs/11 Q-A19）
    line_text: Mapped[str] = mapped_column("text", Text, nullable=False)
    # 版权来源追踪（docs/11 Q-B19）：逐句歌词文本本身可含商用内容，
    # 冗余自 songs.source 便于逐句审计；Java 写须与 songs.source 保持一致
    source: Mapped[str] = mapped_column(
        String(32), nullable=False, server_default=text("'public_domain'")
    )

    __table_args__ = (
        UniqueConstraint("song_id", "seq", name="uq_lrc_song_seq"),
        CheckConstraint("source IN ('public_domain', 'original', 'demo_only')", name="source"),
    )


class SongPitchRef(CreatedAtMixin, Base):
    """参考旋律音高序列（Python 写；docs/06 §9.4 音准评分的「离线预提取」产物）。

    - 与 lrc 逐句 1:1（lrc_id 外键；LRC 重写 → 本表级联删除 → 触发重提取），
      这是「Java 内容库 vs Python 算法产物」耦合点的唯一交点；
    - pitch_ref：该句参考 F0 序列（float 数组，帧 hop 512@16k）；
    - extractor/version：可追溯（pyin 参数或算法升级后重新提取，避免同曲不同标定）。
    """

    __tablename__ = "song_pitch_refs"

    id: Mapped[int] = bigint_pk()
    lrc_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("lrc.id", ondelete="CASCADE"), nullable=False
    )
    start_ms: Mapped[int] = mapped_column(BigInteger, nullable=False)
    end_ms: Mapped[int] = mapped_column(BigInteger, nullable=False)
    pitch_ref: Mapped[dict] = mapped_column(jsonb(), nullable=False)  # {"f0s":[...], "notes":[...]}
    extractor: Mapped[str] = mapped_column(
        String(32), nullable=False, server_default=text("'pyin'")
    )
    version: Mapped[str] = mapped_column(String(16), nullable=False)

    __table_args__ = (UniqueConstraint("lrc_id", name="uq_song_pitch_refs_lrc_id"),)


class ListeningMaterial(TimestampMixin, Base):
    """听力素材（Java 写；docs/06 §9.6「1 个听力素材示例」，推荐候选池第三类）。

    说明：docs/06 §10 表清单未列本表，属设计补充（依据 §9.5/§9.6），待组长拍板；
    若砍掉，推荐候选池回退为「场景+歌曲」两类并在 docs/06 同步。
    """

    __tablename__ = "listening_materials"

    id: Mapped[int] = bigint_pk()
    title: Mapped[str] = mapped_column(String(128), nullable=False)
    level: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    audio_url: Mapped[str] = mapped_column(String(512), nullable=False)
    duration_s: Mapped[int | None] = mapped_column(BigInteger)
    transcript: Mapped[str | None] = mapped_column(Text)  # ASR 字幕示例
    interest_tags: Mapped[dict] = mapped_column(
        jsonb(), nullable=False, server_default=text("'[]'")
    )
    # 版权来源（docs/11 Q-B19）：新闻/音频素材同样有版权，不默认 public_domain
    source: Mapped[str | None] = mapped_column(String(32))
    license: Mapped[str | None] = mapped_column(String(64))
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text(f"'{ContentStatus.DRAFT}'")
    )

    __table_args__ = (
        CheckConstraint("level BETWEEN 1 AND 4", name="level"),
        CheckConstraint(
            "source IN ('public_domain', 'original', 'demo_only') OR source IS NULL",
            name="source",
        ),
        CheckConstraint(
            f"status IN ('{ContentStatus.DRAFT}', '{ContentStatus.PUBLISHED}', "
            f"'{ContentStatus.ARCHIVED}')",
            name="status",
        ),
        Index("ix_listening_materials_status_level", "status", "level"),
    )


class PlacementQuestion(TimestampMixin, Base):
    """入学测试题库（**Java 写**；docs/06 §9.2「admin 题库预置，保证可复现」）。

    - exam_revision：版本号（placements.exam_revision 引用本表版本；改题=新版本，不改历史）；
    - item_index：题序（UNIQUE(exam_revision, item_index)，即 1..n 顺序，不再单设 order_num）；
    - kind：read（固定朗读句）/ qa（1 轮 QA）；reference_answer 供 QA 参考/判定口径；
    - 新增表依据：docs/11 Q-B06（P0）——原设计只有 placements 结果表，题库无处存。
    """

    __tablename__ = "placement_questions"

    id: Mapped[int] = bigint_pk()
    exam_revision: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    item_index: Mapped[int] = mapped_column(SmallInteger, nullable=False)
    kind: Mapped[str] = mapped_column(String(16), nullable=False, server_default=text("'read'"))
    prompt: Mapped[str] = mapped_column(Text, nullable=False)
    reference_answer: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text(f"'{ContentStatus.PUBLISHED}'")
    )

    __table_args__ = (
        UniqueConstraint("exam_revision", "item_index", name="uq_placement_questions_rev_idx"),
        CheckConstraint("kind IN ('read', 'qa')", name="kind"),
        CheckConstraint(
            f"status IN ('{ContentStatus.PUBLISHED}', '{ContentStatus.ARCHIVED}')",
            name="status",
        ),
        Index("ix_placement_questions_revision", "exam_revision"),
    )


class ShadowMaterial(TimestampMixin, Base):
    """影子跟读素材内容库（**Java 写**；local/26 §6 / local/31 §2.4，2026-09-02 设计）。

    - 与 scenarios 并列的候选池第二类；难度先验由 Python 侧 material_difficulty 产出；
    - level：内容方初评 1-4（与 scenarios.difficulty 同语义，仅作难度兜底 FALLBACK_LEVEL）；
    - wpm：原声语速（流利度难度特征）；audio：示范音频（慢速由 TTS/变速派生，不另存大文件）；
    - source 版权红线同 songs（docs/06 §9.7：商用音乐/音频严禁入库）；
    - 人工标注审计（local/32 A-3.3）建议列，随 0003 评估（pending_review 走 material_difficulty）。
    """

    __tablename__ = "shadow_materials"

    id: Mapped[int] = bigint_pk()
    title: Mapped[str] = mapped_column(String(128), nullable=False)
    level: Mapped[int] = mapped_column(SmallInteger, nullable=False)  # 内容方初评 1-4
    text_content: Mapped[str] = mapped_column(Text, nullable=False)  # 跟读文本（逐句，供特征/字幕）
    audio_url: Mapped[str] = mapped_column(String(512), nullable=False)
    wpm: Mapped[int | None] = mapped_column(SmallInteger)  # 原声语速（词/分）
    duration_s: Mapped[int | None] = mapped_column(BigInteger)
    interest_tags: Mapped[dict] = mapped_column(
        jsonb(), nullable=False, server_default=text("'[]'")
    )
    source: Mapped[str] = mapped_column(
        String(32), nullable=False, server_default=text("'public_domain'")
    )
    status: Mapped[str] = mapped_column(
        String(16), nullable=False, server_default=text(f"'{ContentStatus.DRAFT}'")
    )

    __table_args__ = (
        CheckConstraint("level BETWEEN 1 AND 4", name="level"),
        CheckConstraint("source IN ('public_domain', 'original', 'demo_only')", name="source"),
        CheckConstraint(
            f"status IN ('{ContentStatus.DRAFT}', '{ContentStatus.PUBLISHED}', "
            f"'{ContentStatus.ARCHIVED}')",
            name="status",
        ),
        Index("ix_shadow_materials_status_level", "status", "level"),
    )

File: apps/admin/src/api/dto/content.ts
/**
 * 运营域 DTO：内容上下架与工单（docs/50 §6.1 / §10.2、docs/06 §9.6）。
 *
 * ⚠️ 内容行**不是一个形状**。权威来源：`console/content/ConsoleContentController` 的四个
 * record（`SongRow` / `MaterialRow` / `ScenarioRow` / `QuestionRow`）+ `PublishView`，
 * 以及 `console/content/PublishService.PublishResult`。
 *
 * v1 用**单个** `ContentRow {id,title,status,meta,updatedAt}` 覆盖四个域，而后端返回四种形状、
 * 且**没有** `meta` 字段 —— 于是内容页的「元数据」列在运行时恒为空（`row.meta ?? '—'` 永远走 `—`）。
 * 本版按域拆开，页面直接读真实字段。
 */

// ── 运营（内容上下架） ───────────────────────────────────────────────────

export type PublishStatus = 'draft' | 'published' | 'archived'

/** 上下架内容域（`PublishService.DOMAIN_*`）：题库无 draft 故无 publish（docs/50 §4.2） */
export type PublishDomain = 'song' | 'listening' | 'scenario'

/** 上架流水按目标类型过滤（`ConsoleContentController.publishEvents` 的 `targetType` 正则） */
export type PublishTargetType = 'song' | 'listening_material' | 'scenario' | 'book' | 'chapter'

/**
 * 四个内容域的**公共可读字段**。
 *
 * 只放后端确实都有的三个字段：题库的题干是 `prompt` 而不是 `title`（`QuestionRow`），
 * 所以标题列由各域的列定义自己提供 —— 这里刻意不放 `title`，避免再造一个"看着能用、取到 undefined"的字段。
 */
export interface ContentRowBase {
  id: number
  status: PublishStatus | string
  updatedAt: string
}

/** `ConsoleContentController.SongRow`（`GET /content/songs`） */
export interface SongRow extends ContentRowBase {
  title: string
  artist: string | null
  level: number | null
  audioUrl: string | null
  /** 参考旋律状态：`missing|building|ready|invalid`；上架前置校验要求 `ready` */
  pitchRefStatus: string | null
}

/** `ConsoleContentController.MaterialRow`（`GET /content/listening-materials`） */
export interface MaterialRow extends ContentRowBase {
  title: string
  level: number | null
  audioUrl: string | null
  /** 服务端算好的布尔位（`transcript != null && !isBlank`），不是原文 */
  hasTranscript: boolean | null
}

/** `ConsoleContentController.ScenarioRow`（`GET /content/scenarios`） */
export interface ScenarioRow extends ContentRowBase {
  title: string
  sceneType: string | null
  difficulty: number | null
  /** `PublishService.countCorpusItems`：按 `English|中文` 逐行计数（与 Python 解析同口径） */
  corpusItemCount: number
}

/** `ConsoleContentController.QuestionRow`（`GET /content/questions`，只读） */
export interface QuestionRow extends ContentRowBase {
  examRevision: number | null
  itemIndex: number | null
  kind: string | null
  prompt: string | null
}

/**
 * 上架校验失败项（46011 的 `data.violations[]`）。
 * 权威：`PublishService.Violation(field, code, message)` —— v1 写成 `{field, reason}`，
 * 于是字段级原因在弹窗里恒为 `undefined`（会显示成「· lrc：undefined」）。
 */
export interface PublishViolation {
  field: string
  code: string
  message: string
}

/** `ConsoleContentController.PublishView`（`POST /content/{domain}/{id}/publish` 的 data） */
export interface PublishResult {
  domain: string
  id: number
  prevStatus: string | null
  nextStatus: string
  publishedAt: string
}

/**
 * 上架流水一行（`GET /content/publish-events`）。
 *
 * ⚠️ 该端点的 data 不是 record 而是 `LinkedHashMap`（`ConsoleContentController.publishEvents`
 * 明确列出键名），且**没有** `domain` / `prevStatus` / `nextStatus`：
 * - 内容域由 `action`（`content.{domain}.publish`）推导 —— 与后端 `applyPublish` 的写法一致；
 * - 前后状态只存在于 `detail`（`AdminAuditLogEntity.detail` 的原始 JSON 文本）里，故需解析。
 * v1 的 `adminUsername` 真实字段名是 `operator`；时间列是 `createdAt`（后端**没有** `publishedAt`）。
 */
export interface PublishEventRow {
  id: number
  /** 形如 `content.song.publish` */
  action: string
  targetType: string
  targetId: string
  /** 操作人账号名（`admin_audit_logs.admin_username` 快照） */
  operator: string
  adminUserId: number | null
  summary: string | null
  /** 未解析的 JSON 文本；前后状态经 `publishEventDetail()` 取 */
  detail: string | null
  createdAt: string
}

/** 工单状态机（docs/06 §9.6）：open → processing → resolved → closed，**禁回退**、closed 终态 */
export type TicketStatus = 'open' | 'processing' | 'resolved' | 'closed'

/** 工单类型：反馈 / 报错 / 内容纠误（content_correction 时才有 targetType/targetId） */
export type TicketKind = 'feedback' | 'bug' | 'content_correction'

// ── Python 写方：书籍 / 章节 / 媒体（`/api/v1/console/library/**`） ────────
// 权威：`services/python/app/console/api/routes/library.py` 的
// `_book_view`(:285) / `_chapter_view`(:298) / `_media_view`(:311)。
// 书籍与媒体由 Python 服务写（单写方矩阵），故字段名同为 snake_case。

export type BookLevel = 'L1' | 'L2' | 'L3' | 'L4'

/** 书籍列表行（library.py:285-295） */
export interface LibraryBookRow {
  id: number
  title: string
  author: string
  level: BookLevel
  status: PublishStatus
  /** 书的**全部**章节数（seed 派生计数）——列表响应**没有**"已上架章节数" */
  chapter_count: number
  word_count: number
  updated_at: string | null
}

/** 章节行（library.py:298-308） */
export interface LibraryChapterRow {
  id: number
  book_id: number
  chapter_no: number
  title: string
  status: PublishStatus
  word_count: number
  char_count: number
  updated_at: string | null
}

/** 媒体类型：`image|video|avatar`（models/media.py:62 的 CHECK；**没有 audio**） */
export type MediaKind = 'image' | 'video' | 'avatar'
/** 媒体状态（models/media.py:65 的 CHECK） */
export type MediaStatus = 'ready' | 'hidden' | 'deleted'

/**
 * 媒体行（library.py:311-322）。
 * ⚠️ 后端**不返回** `width`/`height`/`duration_s`（DB 里有列，但 `_media_view` 未投影），
 * 因此尺寸与时长在控制台无数据可显示，页面不得再读它们。
 */
export interface MediaAssetRow {
  /** 后端把 public_id 同时塞进 id 与 public_id（library.py:313-314） */
  id: string
  public_id: string
  owner_id: number
  kind: MediaKind
  mime_type: string
  size_bytes: number
  status: MediaStatus
  /** 公共读取地址 `"/api/v1/media/{public_id}"`（隐藏后该端点即读不到） */
  url: string
  created_at: string | null
}

// ── 内容写入（运营的增改删） ─────────────────────────────────────────────
//
// 权威：`console/content/ConsoleContentWriteController` 的六个 record 与四个 `xView`。
// ⚠️ 两个容易踩的形状差异：
// 1. **列表行 ≠ 可编辑实体**。`GET /content/songs` 的 `SongRowList` 只有 6 个字段，
//    而表单要的 `durationS`/`bpm`/`musicalKey`/`lrcUrl`/`coverUrl` 只在**单条读取**
//    （`GET /content/songs/{id}` 的 `songView`）里才有 —— 所以编辑弹窗必须按 id 回读，
//    不能拿列表行凑（那会把没显示的字段清空）。
// 2. **更新是 PUT 而不是 PATCH**，且请求体是**全量** `xUpsert`：漏字段 = 服务端按
//    `applySong` 的默认值覆盖（`interestTags→"[]"`、`source→public_domain`、
//    `status→draft`、`pitchRefStatus→missing`）。因此表单必须把 status 显式回填，
//    否则"编辑一次歌曲"会把已上架的歌曲悄悄打回草稿。

/** 内容审核态（`PublishService.STATUS_*`）；题库的取值域只有 publish/archive 两态 */
export type ContentStatus = 'draft' | 'published' | 'archived'

/** 歌曲版权来源（`SongUpsert.source` 的 `@Pattern`） */
export type SongSource = 'public_domain' | 'original' | 'demo_only'

/** 参考旋律状态（`SongEntity.pitchRefStatus`，Python 离线任务写、Java 只读） */
export type PitchRefStatus = 'missing' | 'building' | 'ready' | 'invalid'

/** 场景类型（`ScenarioUpsert.sceneType` 的 `@Pattern`，是**封闭**取值域，不是自由文本） */
export type SceneType = 'cafe' | 'airport' | 'interview' | 'library' | 'other'

/** 题目类型（`QuestionUpsert.kind` 的 `@Pattern`）：read = 朗读题、qa = 问答题 */
export type QuestionKind = 'read' | 'qa'

/**
 * 歌曲**单条**视图（`ConsoleContentWriteController.songView`）。
 * 与列表 `SongRowList` 是两个形状 —— 这里刻意不复用，避免"取到 undefined 还以为是空值"。
 */
export interface SongDetail extends ContentRowBase {
  title: string
  artist: string | null
  level: number | null
  durationS: number | null
  bpm: number | null
  musicalKey: string | null
  audioUrl: string | null
  lrcUrl: string | null
  coverUrl: string | null
  interestTags: string | null
  source: string | null
  pitchRefStatus: string | null
}

/** 场景单条视图（`scenarioView`） */
export interface ScenarioDetail extends ContentRowBase {
  title: string
  sceneType: string | null
  difficulty: number | null
  description: string | null
  systemPrompt: string | null
  openingLine: string | null
  targetCorpus: string | null
  interestTags: string | null
  promptVersion: number | null
  estimatedTurns: number | null
  estimatedMinutes: number | null
}

/** 听力素材单条视图（`materialView`）；注意这里**有** transcript 原文（列表行只有布尔位） */
export interface MaterialDetail extends ContentRowBase {
  title: string
  level: number | null
  audioUrl: string | null
  durationS: number | null
  transcript: string | null
  interestTags: string | null
  source: string | null
  license: string | null
}

/** 题目单条视图（`questionView`） */
export interface QuestionDetail extends ContentRowBase {
  examRevision: number
  itemIndex: number
  kind: string
  prompt: string
  referenceAnswer: string | null
}

/**
 * LRC 单行（`GET|PUT /content/songs/{id}/lrc` 的元素）。
 * `seq` 由服务端按请求数组顺序重排（`replaceLrc`），前端**不回传** seq。
 */
export interface LrcLineRow {
  seq: number
  offsetMs: number
  endOffsetMs: number | null
  lineText: string
  /** 继承 `songs.source`（docs/11 Q-B19 冗余一致性），前端只读 */
  source: string | null
}

/** LRC 行写入体（`ConsoleContentWriteController.LrcLine`） */
export interface LrcLineInput {
  offsetMs: number
  endOffsetMs: number | null
  lineText: string
}

/** LRC 整首重写体（`LrcUpsert`） */
export interface LrcUpsert {
  lines: LrcLineInput[]
}

/** 歌曲写入体（`SongUpsert`）。`pitchRefStatus` 不在前端表单里，见 `contentFormPayload` 的说明。 */
export interface SongUpsert {
  title: string
  artist: string | null
  level: number
  durationS: number | null
  bpm: number | null
  musicalKey: string | null
  audioUrl: string
  lrcUrl: string | null
  coverUrl: string | null
  interestTags: string | null
  source: SongSource | null
  status: ContentStatus | null
}

/** 场景写入体（`ScenarioUpsert`） */
export interface ScenarioUpsert {
  title: string
  sceneType: SceneType
  difficulty: number
  description: string | null
  systemPrompt: string
  openingLine: string
  targetCorpus: string | null
  interestTags: string | null
  promptVersion: number | null
  estimatedTurns: number | null
  estimatedMinutes: number | null
  status: ContentStatus | null
}

/** 听力素材写入体（`MaterialUpsert`） */
export interface MaterialUpsert {
  title: string
  level: number
  audioUrl: string
  durationS: number | null
  transcript: string | null
  interestTags: string | null
  source: SongSource | null
  license: string | null
  status: ContentStatus | null
}

/** 题目**新建**体（`QuestionUpsert`）：版本号与序号只在新建时可定 */
export interface QuestionUpsert {
  examRevision: number
  itemIndex: number
  kind: QuestionKind
  prompt: string
  referenceAnswer: string | null
  status: 'published' | 'archived' | null
}

/**
 * 题目**编辑**体（`QuestionPatch`，`PUT /content/questions/{id}`）。
 *
 * ⚠️ 后端**不允许**改 `examRevision`/`itemIndex` —— 它们是题目身份（docs/10 §3.2：
 * 改题=新版本，不改历史）。所以编辑表单里这两项只读回显。
 */
export interface QuestionPatch {
  prompt: string
  referenceAnswer: string | null
  status: 'published' | 'archived' | null
}

/**
 * 工单行。
 * ⚠️ 字段与 Java `TicketView` 逐字对齐（`services/java/.../ticket/dto/TicketView.java`）——
 * v1 的 `{id,userId,kind,status,subject,...}` 是**臆造**的（没有 `subject`，有 `title`/`content`/`adminReply`），
 * 这类"看着合理但后端没有"的字段会让页面在运行时静默空白。
 */
export interface TicketRow {
  id: number
  userId: number
  kind: TicketKind | string
  targetType: string | null
  targetId: number | null
  title: string | null
  content: string
  status: TicketStatus | string
  adminId: number | null
  adminReply: string | null
  resolvedAt: string | null
  createdAt: string
  updatedAt: string
}
